### PR TITLE
Add intro_concurrency talk.

### PR DESCRIPTION
**Pull request by btracey**
This talk was delivered internally at Stanford University by Brendan Tracey

### CODE REVIEW

**Comment by jonlawlor** [comment]
The charts comparing languages remind me very much of Martin Odersky's talk "The Trouble with Types," starting around 10 minutes in.

**Comment by btracey** [comment]
So they do. As denoted in the slide, they should most remind you of http://talks.golang.org/2014/gocon-tokyo.slide#16

File: 2015/intro_concurrency.slide [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+- Statically-Typed
+- Garbage collected
+- Concurrent
+- My language of choice
+
+* Origins
+Designed at Google by 
+
+- Ken Thompson (C, Unix)
+- Rob Pike (UTF-8)
```
Comment (kortschak) [inline 2015/intro_concurrency.slide]: This is diminishing Rob; add Plan9 and Inferno, Squeak and NewSqueak? Also note that UTF-8 was him and Ken.

File: 2015/intro_concurrency.slide [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+- No makefiles (go install)
+- Testing suite (go test)
+- Race/Deadlock detection (go install -race)
+- Automatic documentation (godoc)
+- Code formatting (gofmt)
+- Automatic import detection (goimports)
+- Profiling (go tool pprof)
+
+* Why Go
+Better than any other language to (personal opinion):
```
Comment (kortschak) [inline 2015/intro_concurrency.slide]: s/to/too/

File: 2015/intro_concurrency.slide (hunk 1 above)
Comment (btracey) [inline 2015/intro_concurrency.slide]: Taking out the accolades may just be easier. I was giving examples my audience would be familiar with (not attempting to diminish).

Comment (btracey) [inline 2015/intro_concurrency.slide]: I could add to the lists if you prefer (even if they will be unfamilar to many numeric programmers)

File: 2015/intro_concurrency.slide [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+.play intro_concurrency/interface.go /MonteCarlo estimates/,$
+
+* Concurrency
+
+* Concurrency
+
+MPI:
+
+- One source code, but multiple instances running
+- One thread per instance
```
Comment (kortschak) [inline 2015/intro_concurrency.slide]: Is this true? I guess it depends on the definition of thread, but I could imagine an MPI program with workers each with a number of system threads. Just curious really.

File: 2015/intro_concurrency.slide [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+* Goroutines
+The `go` command launches a concurrently executing process
+
+  go func()
+  go func(x, y, ...)
+
+Like a thread, but very cheap
+
+- Goroutines execute independently from one another
+- Millions of goroutines common
```
Comment (kortschak) [inline 2015/intro_concurrency.slide]: ... in some domains.

File: 2015/intro_concurrency.slide (hunk 3 above)
Comment (btracey) [inline 2015/intro_concurrency.slide]: I believe the real answer to this question is that the set of things you can do with MPI is much larger than the set of things that's most frequently done with MPI. 

I can put a (frequently) at the end of the bullet.

File: 2015/intro_concurrency.slide (hunk 2 above)
Comment (btracey) [inline 2015/intro_concurrency.slide]: Not a correct substitution. I'll change it to make the wording less awkward.

File: 2015/intro_concurrency.slide [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+
+- Experimental
+- Not yet as fast or robust as traditional MPI libraries (uses TCP for communication, some bugs)
+- No special compiler, very simple API 
+    mpi.Send(data interface{}, rank, tag int)
+    mpi.Receive(data interface{}, rank, tag int)
+- Trivial to incorporate shared memory (one process per node instead of per CPU)
+- Implementation about 600 lines of code
+
+DEMO (github.com/btracey/mpi/examples/helloworld)
```
Comment (btracey) [inline 2015/intro_concurrency.slide:363]: I'm happy to remove this slide if you don't think it really fits in. It's there to highlight what's possible, but I don't expect my implementation to become a standard or anything.

**Comment by kortschak** [comment]
Did you really put R in the fun category?

File: 2015/intro_concurrency.slide (hunk 2 above)
Comment (kortschak) [inline 2015/intro_concurrency.slide]: Ahh, sorry.

**Comment by btracey** [comment]
I've never used R. It's the most polarizing language in my experience. Some love it, others don't at all.

File: 2015/intro_concurrency.slide (hunk 1 above)
Comment (kortschak) [inline 2015/intro_concurrency.slide]: The reason for adding at least Plan9 and the squeaks is that the thinking that went into them has contributed to how Go is now AFAICS.

File: 2015/intro_concurrency.slide (hunk 5 above)
Comment (kortschak) [inline 2015/intro_concurrency.slide:363]: I think it should stay.

**Comment by kortschak** [comment]
R is a nightmare. I use it when I want nice graphics and when I want to do stats reasonably fast, but it is opaque and surprising; the source not immediately available except for the R code (the C/Fortran must be dug for and the calling is sometimes baroque), the typing is truly bizarre and impacts on the control flow often in disturbing ways because of arbitrary author choices and there is no community consistency in approach so that data from one package must often be significantly reshaped/reorganised to by used with another.

File: 2015/intro_concurrency.slide (hunk 3 above)
Comment (kortschak) [inline 2015/intro_concurrency.slide]: I'm not fussed, I was just curious.

**Comment by btracey** [comment]
I could make the R circle very tall? I'm not interested in picking a fight with the R crowd (I'll leave that to you).

File: 2015/intro_concurrency.slide [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+
+.play intro_concurrency/chan_range.go /func consumer/,$
+
+* Deadlock Detection (not perfect)
+
+.play intro_concurrency/deadlock.go
+
+Race detector available as well
+  
+  go install -race myprogram.go
```
Comment (kortschak) [inline 2015/intro_concurrency.slide]: I thought I commented this here, but obviously not.

It's my understanding that you can do a race detector run in a slide by holding shift when you run the code, but I can't get it to work.

**Comment by kortschak** [comment]
> I could make the R circle very tall? I'm not interested in picking a fight with the R crowd (I'll leave that to you).

It's fine how it is.

**Comment by kortschak** [comment]
LGTM

Comments can be taken or left as desired.

**Comment by kortschak** [comment]
BTW How was it received?

File: 2015/intro_concurrency.slide (hunk 6 above)
Comment (btracey) [inline 2015/intro_concurrency.slide]: Works on my machine. I'll add a slide.

Is this behavior documented anywhere? I thought something like that would be available, but couldn't find anything.

**Comment by btracey** [comment]
It was reasonably well received. Most of the audience are developers/users of large established codebases in python and C++. Those packages aren't going to be re-written, and so there aren't a bunch of people wanting to jump ship. The talk seemed to pique the interest of a few who are "tempted to try it out". 

Many around me are doing HPC kind of things, so 
1) The lack of robust support for distributed memory computation is a problem
2) They really want to see performance numbers, which is reasonable.

In general, I think they are impressed by the easy syntax, and could get excited if it lives up to the "almost C" claims.

**Comment by btracey** [comment]
Is there any easy way to show this to people (like talks.golang.org), or do I just add that to the list of reasons why I need to build a website?

File: 2015/intro_concurrency.slide (hunk 4 above)
Comment (btracey) [inline 2015/intro_concurrency.slide]: I did it once when I was really lazy with a monte carlo simulation...

**Comment by btracey** [comment]
Thanks for the comments

**Comment by btracey** [comment]
Thanks for answering the question about the website. Just had to dig some more.

https://groups.google.com/d/msg/golang-nuts/u9z1YQ8GV_8/G3IOkETnj-QJ

**Comment by kortschak** [comment]
Good, you've found talks.godoc.org.

**Comment by btracey** [comment]
Do you know if each code has a local GOPATH? That way I could vendor ExtendedRosenbrock and have all the examples work. I tried to find it in the source code, but did not succeed.

File: 2015/intro_concurrency.slide [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+.play intro_concurrency/mc.go /func main/,$
+
+* Interfaces
+An *interface* is a set of methods
+
+  type Distribution interface {
+      Rand() float64
+  }
+
+A type *satisfies* the inteface if it has those methods
```
Comment (kortschak) [inline 2015/intro_concurrency.slide:165]: Sorry, missed this one: s/inteface/interface/

**Comment by kortschak** [comment]
You won't be able to play some of your code because of imports.

**Comment by btracey** [comment]
That's too bad. I was hoping if I added the correct file structure it might be okay.

**Comment by kortschak** [comment]
You could simulate it by having a type that incorporated all the package contents and then you make a value named after the package you are simulating. Seems like a lot of work though. In my talks I have a dummy go file with imports of all the packages that are required; then at least a go get will pull all the dependencies.

**Comment by btracey** [comment]
```
You could simulate it by having a type that incorporated all the package contents and then you make a value named after the package you are simulating
```

That's amazingly wicked. I love it. 

What do you mean by a "dummy go file"? This is at the level of name.slide?

**Comment by kortschak** [comment]
> That's amazingly wicked. I love it.

I used to be a perl programmer.

Don't do it for generally published talks (where people see the code). It's horrible really. I do a similar thing that is less underhanded in my lecture slides so I can alter behaviour of the packages I import from their normal operation.

For more deviousness, see [this discussion](https://groups.google.com/d/topic/golang-nuts/fSz4iI9y2O4/discussion), leading to this: http://play.golang.org/p/CXT6KuLAp1 (the shorted bash version is not portable to windows).

> What do you mean by a "dummy go file"? This is at the level of name.slide?

Here it would be .../intro_concurrency/imports.go with an import statement block doing blank imports for all the non-standard library imports you make in snippets. This way, when you `go get github.com/gonum/talks/2015/intro_concurrency` you get all the dependencies.

**Pull request merged**